### DISCUSSION TITLE
Regarding v4

### DISCUSSION THREAD

**Discussion by thecodejack** [data-driven-forms/react-forms#1015] [General]
Hi Team

I see that there is some active development happening for v4 and it seems to be removing `final-form`. Just wanted to know if there will be any planned major breaking changes?

**Comment by rvsia** [comment]
Hello, yeah, our goal is to remove the `final-form`, because there are some issues and the author ignores them. Our framework will be better suited for React and there will be some enhancements over FF. And with more control we can provide more features in the future.

Also, we plan to make the API the same so we do not expect **any breaking changes** for users that are using DDF according to our documentation. For users, that are using some undocumented (on our page) FF components/mutators, there could be breaking changes but we can work on them.

**Reply by rvsia** [reply]
(However, maybe we will use the release as an opportunity to introduce some unrelated breaking changes. It depends how quick we will be with the implementation and testing of it. Also, we are discussing some ability to switch between FF/our state manager. In case that v4 is released sooner than expected, we will still support v3 for some time as it is still a new release.)

**Reply by Hyperkid123** [reply]
The v4 is not going to happen for a while. The state manager is at the core of everything so we have to make sure nothing is broken. Sire, our current test cases are passing but we have to start testing the new state manager on real examples. So there will be probably some opt-in flag to test it out early or something like that. We don't know yet. But in our internal testing, we use v3 mappers v3 renderer and schemas used for v3 testing. So it looks promissing so far.

As Richard said, we are not changing any of the API. We have decided on this early on to prevent lengthy migration. We will keep the same API as FF has right now. Sure we can get accused that we are just FF copy cat, but it's better than migrating existing forms. We ourselves have a lot of them.

**Reply by thecodejack** [reply]
Thanks for quick reply. Also wondering if there are any performance benefits you are expecting with V4?

coz I see there is some performance degradation if a simple form becomes huge.

https://codesandbox.io/s/poc-performance-efxhm?file=/src/App.js

**Reply by Hyperkid123** [reply]
We have tested the v4 even on larger forms and it is smooth with the correct subscription. Sure first render will take longer, but there is no way around that.

However, it should OK even with the final form manager. I'll check out your example, probably later today, and see what might be the issue. Seems like the subscription does not work in your example.

But form with 350 fields, that might be a bit too much for the renderer with conditions and form spies etc. We will see.  You might have found a space for improvement.

**Reply by thecodejack** [reply]
Initially rendering is fine to get delayed in my case. My concern is when user types text, it taking bit too much time. The `counter` in codesandbox link is `50`. I was initially trying with `100` when delay is more. As you said, probably coz of spies. Would love to know if there is any work around to improve same. Thanks.

**Reply by rvsia** [reply]
I think a huge form like this one with all the functionality (condititions etc.) is going to have performance issues no matter what.

However, because there is no reason to show all the fields at once, I would make a component that will show just a static version of the form section and users should have to click to “enable it”.. or make some kind of virtualization and only render items that users can see. Or to use pagination. Etc.

**Reply by Hyperkid123** [reply]
> Initially rendering is fine to get delayed in my case. My concern is when user types text, it taking bit too much time. The `counter` in codesandbox link is `50`. I was initially trying with `100` when delay is more. As you said, probably coz of spies. Would love to know if there is any work around to improve same. Thanks.

@thecodejack I apologize I had no time to debug this yesterday. I may have time later today, although a lot of work just came up.

**Reply by thecodejack** [reply]
> I think a huge form like this one with all the functionality (condititions etc.) is going to have performance issues no matter what.

Agreed. I am just trying to find limits of DDF. 


> static version of the form section and users should have to click to “enable it”.. or make some kind of virtualization and only render items that users can see. Or to use pagination. Etc.

Or how will it perform if I break this into multiple DDF forms?

**Reply by Hyperkid123** [reply]
The limiting factor is just "how much stuff is on the screen". So for example, if you have a wizard-like form that only shows let's say 10 fields at the same time, but the whole wizard has 100 fields, you would still only render 10 items which obviously is much cheaper than 100.

If you want to keep all items on the screen at the same time, the quickest sollution is using `React.memo` between `useFieldApi` and the actual component (e.g. Mui TextField) to save some resources.

BUT! This is a profiler sample from your form on keypress in an input field:
![Screenshot from 2021-03-24 12-25-07](https://user-images.githubusercontent.com/22619452/112303080-2b0d2280-8c9c-11eb-8a71-4c85827a7a77.png)

As you can see there are 28 renders (some of them are empty) and one (the yellow one) renders every single field in your form . That is the point that causes the lag. On its own, it takes almost a second. So the lag is not caused by the state manager, but mostly by the render time.

So the `React.memo` will help with this. But this should not be happening as by default, we do not subscribe to value change. I think this problem is caused by the `Condition` wrapper and the fact that it's not memoized by the condition result. We should definitely check if memorizing the condition will help with this issue, or if it's a subscription issue in the field array component.

Unfortunately, I am super busy right now. Maybe tomorrow or on Friday I might find some time to check it out. Or if you want to take a stab at it @thecodejack feel free.

EDIT: I would just like to point out that there are not 28 renders on input field change, I have focused multiple elements, and renders contain focus/blur events. Also, every renders except for the long one took just a few ms, so that is why I claim the rendering is the problem, not the state management and other computations.

**Reply by thecodejack** [reply]
So here is what I tried....

1.  With current setup in codesandbox

First Keypress: ~2 secs
Second Key Press and later: ~ 160ms


2. Made own TextField...added a React.Memo to TextField. Commented out `Condition`...

First Key press: ~ 2 secs
Second Key Press and later: ~ 83ms

**Improvement observed in later presses**


3. UnCommented `Condition`

Same metrics as previous one


Link https://codesandbox.io/s/poc-performance-efxhm?file=/src/App.js

I did observe what you said about Performance. `React Final Form` is not killing time. There are multiple React unnecessary triggers which can someway be optimized. I will try spending more time how these unnecessary renders can be avoided.

**Reply by Hyperkid123** [reply]
Looking at the example, the memo is misplaced. It should be something like this:
```jsx

const MemoizedTextField = memo((props => {
  return <MuiTextField />
}), (prevProps, nextProps) => /** props comparator if necessary, react is doing only shallow equal and wotn work on functions aswell */)

const TextField = (props) => {
  const {input, meta, ...rest} = useFieldApi(props)
  return <MemoizedTextField {...} />
}
```

The reason why the first render happens because the `pristine` global flag is changed. Which triggers the `useFieldApi` hook and we have a render. The memo can shield you from that. Other renders as well. If we configure the comparator properly. But we should add some caching to the renderer as well.

**Reply by thecodejack** [reply]
Thanks. Followed your suggestion. I removed all fields which are not TextField. 

1. And tried rendering 500*2 TextFields

**Observed 3 secs lag for first keypress. 2nd and 3rd are 35ms**

2. Used memoized TextField. Also made TextField memoized for all props except value. 

**Observed 500 ms lag for first keypress. 2nd and 3rd are 35ms.**


**Althgh 500ms lag can still be observed, it's far better improvement.**  🎉 


Tried drilling down a bit on perf. With both Chome FlameGraph and React FlameGraph. I can see that All textfields are being triggered for render. Due to memoization in 2nd, TextFields will not be rendering except. Ideal case would, there is no render trigger for the TextFields where we are not entering data. I mean initially was under impression that it's not `Final Form`. But looks like changes to data from `Final Form` might be causing trigger by react.

**Reply by rvsia** [reply]
I think you can remove the first long render by reverting all defaul subscripton values to false: https://github.com/data-driven-forms/react-forms/blob/68873fac7183dbeb3baea61b0d9b68d8e8e3424f/packages/react-form-renderer/src/form-renderer/form-renderer.js#L58

**Reply by Hyperkid123** [reply]
The `pristine` flag is the main cause for the initial lag. Any change that shifts the form flag `pristine`  from `true` to `false` will trigger a re-render in all fields. I hope today I will have some time to check the internals and see why we trigger these renders. :crossed_fingers:

**Reply by Hyperkid123** [reply]
@thecodejack so by adding this prop `subscribtion={{ pristine: false }}` we get rid of the initial 500ms lag. The experience is still not super smooth, but we have 1000 fields in the form so we can expect any miracles.

There is one bottleneck in ddf which we can fix. Take a look at following profiler snapshot:
![Screenshot from 2021-03-25 13-56-07](https://user-images.githubusercontent.com/22619452/112476682-742fa600-8d72-11eb-9ea5-8d094415228a.png)

We have a lot of grey "leaves" if you will. The MUI text field si properly memoized and not renderer. That is good. However if we zoom a bit we can see that there is not ay many "gray" candles as we would like:
![Screenshot from 2021-03-25 13-56-51](https://user-images.githubusercontent.com/22619452/112476885-aa6d2580-8d72-11eb-98b9-52a2aaf136b0.png)

As you can see, we still render some internal DDF components, even though it's not required. We have extra renders in `SingleField`, `FormConditionWrapper`, `FormFieldHideWrapper` and the root unmemoized `TextField`. yes each of them takes about 0.1ms 0.2ms to render, but when we multiply it by 1000, we have roughly 30ms of extra render time. I will create an issue and PR and we shall see if we can get it down even further and make the experience smooth. If that won't help I think we have reached the limit of the state management.

**Reply by thecodejack** [reply]
Awesome analysis. Thanks for looking into that. 

Just wanted to know if you are planning to fix it with v4 State setup or v3's `final-form` state manager?

**Reply by Hyperkid123** [reply]
In V3 final form. This is renderer issue which does not change in v4.

**Reply by Hyperkid123** [reply]
https://github.com/data-driven-forms/react-forms/issues/1016 Here is an issue where we will track the progress. Start working on it today.

**Reply by Hyperkid123** [reply]
@thecodejack the performance optimizations have been released in version `3.3.2`. Feel free to test it and see if the performance in your original use case has improved.
